src/ape/api/contracts.py

```diff
@@ -0,0 +1,259 @@
+
+
+@dataclass
+class ContractCall:
+    abi: ABI
+    provider: ProviderAPI
+
+    def __repr__(self) -> str:
+        return self.abi.signature
+
+    def encode(self, *args, **kwargs) -> TransactionAPI:
+        return self.provider.network.ecosystem.encode_transaction(self.abi, *args, **kwargs)
+
+    def __call__(self, *args, **kwargs) -> ReceiptAPI:
+        txn = self.encode(*args, **kwargs)
+
+        if "sender" in kwargs and not isinstance(kwargs["sender"], str):
+            txn.sender = kwargs["sender"].address
+
+        return self.provider.send_call(txn)
+
+
+@dataclass
+class ContractCallHandler:
+    provider: ProviderAPI
+    abis: List[ABI]
+
+    def __call__(self, *args, **kwargs) -> ReceiptAPI:
+        # NOTE: Should only match one ABI in set
+        abi = next(abi for abi in self.abis if len(args) == len(abi.inputs))
+        return ContractCall(  # type: ignore
+            abi=abi,
+            provider=self.provider,
+        )(*args, **kwargs)
+
+
+@dataclass
+class ContractTransaction:
+    abi: ABI
+    provider: ProviderAPI
+
+    def __repr__(self) -> str:
+        return self.abi.signature
+
+    def encode(self, *args, **kwargs) -> TransactionAPI:
+        return self.provider.network.ecosystem.encode_transaction(self.abi, *args, **kwargs)
+
+    def __call__(self, *args, **kwargs) -> ReceiptAPI:
+
+        if "sender" in kwargs:
+            sender = kwargs["sender"]
+            kwargs["sender"] = sender.address
+
+            txn = self.encode(*args, **kwargs)
+            return sender.call(txn)
+
+        else:
+            raise  # Must specify a `sender`
+
+
+@dataclass
+class ContractTransactionHandler:
+    provider: ProviderAPI
+    abis: List[ABI]
+
+    def __call__(self, *args, **kwargs) -> ReceiptAPI:
+        # NOTE: Should only match one ABI in set
+        abi = next(abi for abi in self.abis if len(args) == len(abi.inputs))
+        return ContractTransaction(  # type: ignore
+            abi=abi,
+            provider=self.provider,
+        )(*args, **kwargs)
+
+
+@dataclass
+class ContractEvent:
+    name: str
+    inputs: List[dict]
+    provider: ProviderAPI
+
+
+@dataclass
+class ContractLog:
+    name: str
+    data: Dict[str, Any]
+
+
+class ContractInstance(AddressAPI):
+    _address: str
+    _contract_type: ContractType
+
+    def __repr__(self) -> str:
+        return f"<{self._contract_type.contractName} {self.address}>"
+
+    @property
+    def address(self) -> str:
+        return self._address
+
+    def __getattr__(self, attr_name: str) -> Any:
+        def get_name(abi: ABI):
+            return abi.name
+
+        try:
+            if attr_name in map(get_name, self._contract_type.events):
+                selected_abis = [
+                    abi for abi in self._contract_type.events if get_name(abi) == attr_name
+                ]
+                assert len(selected_abis) == 1
+                return self._event_class(  # type: ignore
```

Comment:
The `self._event_class` property was part of the old implementation.

Reply:
Also these three blocks are very similar, so you could generalize that inner function to handle three different cases, something like this:

```
def __getattr__(self, attr_name: str) -> Any:
    handlers = {
        'events': EventHandler,
        'calls': ContractCallHandler,
        'transactions': ContractTransactionHandler,
    }
    def get_abi(abi_type -> str, abi_handler -> Any):
        selected = [abi for abi in getattr(self._contract_type, abi_type) if abi.name == attr_name]
        if not selected:
            return
        kwargs = {'provider': self.provider, 'abis': selected[0] if abi_type == "events" else selected}
        assert len(selected) == 1 if abi_type == "events" else True
        return handlers[abi_type](**kwargs)
    for (abi_type, abi_handler) in handlers.items():
        handler = get_abi(abi_type, abi_handler)
        if handler:
            return handler
```

It might only be worth it to do that though if there are more than 3 cases.

Reply:
refactored in 8afd46